contracts/trading/seaport16/ImmutableSeaport.sol

```diff
@@ -0,0 +1,633 @@
+     *
+     * @return The name of this contract.
+     */
+    function _name() internal pure override returns (string memory) {
+        // Return the name of the contract.
+        return "ImmutableSeaport";
+    }
+
+    /**
+     * @dev Internal pure function to retrieve the name of this contract as a
+     *      string that will be used to derive the name hash in the constructor.
+     *
+     * @return The name of this contract as a string.
+     */
+    // slither-disable-next-line dead-code
+    function _nameString() internal pure override returns (string memory) {
+        // Return the name of the contract.
+        return "ImmutableSeaport";
+    }
+
+    /**
+     * @dev Helper function to revert any fulfillment that has an invalid zone
+     */
+    function _rejectIfZoneInvalid(address zone) internal view {
+        if (!allowedZones[zone]) {
+            revert InvalidZone(zone);
+        }
+    }
+
+    /**
+     * @notice Fulfill an order offering an ERC20, ERC721, or ERC1155 item by
+     *         supplying Ether (or other native tokens), ERC20 tokens, an ERC721
+     *         item, or an ERC1155 item as consideration. Six permutations are
+     *         supported: Native token to ERC721, Native token to ERC1155, ERC20
+     *         to ERC721, ERC20 to ERC1155, ERC721 to ERC20, and ERC1155 to
+     *         ERC20 (with native tokens supplied as msg.value). For an order to
+     *         be eligible for fulfillment via this method, it must contain a
+     *         single offer item (though that item may have a greater amount if
+     *         the item is not an ERC721). An arbitrary number of "additional
+     *         recipients" may also be supplied which will each receive native
+     *         tokens or ERC20 items from the fulfiller as consideration. Refer
+     *         to the documentation for a more comprehensive summary of how to
+     *         utilize this method and what orders are compatible with it.
+     *
+     * @param parameters Additional information on the fulfilled order. Note
+     *                   that the offerer and the fulfiller must first approve
+     *                   this contract (or their chosen conduit if indicated)
+     *                   before any tokens can be transferred. Also note that
+     *                   contract recipients of ERC1155 consideration items must
+     *                   implement `onERC1155Received` to receive those items.
+     *
+     * @return fulfilled A boolean indicating whether the order has been
+     *                   successfully fulfilled.
+     */
+    function fulfillBasicOrder(
+        BasicOrderParameters calldata parameters
+    ) public payable virtual override returns (bool fulfilled) {
+        // All restricted orders are captured using this method
+        if (uint256(parameters.basicOrderType) % 4 != 2 && uint256(parameters.basicOrderType) % 4 != 3) {
+            revert OrderNotRestricted();
+        }
```

Comment:
nit: should we make this a bit more legible? maybe extract it to a meaningful function, given it is repeated a few times and add a comment along the lines of 

> // Basic order types (modulo 4): 0 = FULL_OPEN, 1 = PARTIAL_OPEN, 2 = FULL_RESTRICTED, 3 = PARTIAL_RESTRICTED. Only restricted orders (types 2 and 3) are allowed